[WIP] Removing Asset

majecty opened the pull request.

majecty commented on state/src/item/metadata.rs:
```diff
@@ -110,23 +77,24 @@ impl Metadata {
+    // FIXME: I'm not sure of it.
     fn is_null(&self) -> bool {
-        self.number_of_shards == 0
+        self.hashes.is_empty() && self.seq == 0
```
@sgkim126 IMO this code works find. What do you think about it?

sgkim126 replied:
I think the code is right except we should remove hashes too to remove shards.
However, IMO, we've better keep shards.

sgkim126 replied:
FYI, `is_null` should be `true` only when the value means that it does not exist in the trie.

majecty commented:
There are two failed e2e tests.

```
  1) Memory pool memory limit test
       To others
         More than limit:

      AssertionError: expected 5 to equal 0
      + expected - actual

      -5
      +0

      at Context.<anonymous> (src/e2e.long/mempool.test.ts:206:64)
      at processTicksAndRejections (internal/process/task_queues.js:85:5)

  2) sync 2 nodes
       2 nodes
         A-B diverged with the same transaction
           Both transaction success
             "before each" hook for "It should be synced when A-B connected":

      AssertionError: expected { Object (value) } to not deeply equal { Object (value) }
      + expected - actual


      at Context.<anonymous> (src/e2e.long/sync2.test.ts:170:72)
      at processTicksAndRejections (internal/process/task_queues.js:85:5)
```

Remagpie commented on core/src/miner/miner.rs:
```diff
@@ -534,7 +482,7 @@ impl Miner {
         let mut invalid_tx_users = HashSet::new();
 
-        let immune_users = self.immune_users.read();
+        let _immune_users = self.immune_users.read();
```
I think we can remove this.

Remagpie commented on core/src/scheme/scheme.rs:
```diff
@@ -133,6 +134,21 @@ impl Scheme {
+
+        Ok((db, root))
+    }
+
```
I suggest moving this change to a separated commit.
I think it's too important to be aggregated with removing tracket and vm.

Remagpie commented on types/src/errors/syntax_error.rs:
```diff
@@ -77,7 +62,6 @@ const ERROR_ID_TEXT_CONTENT_TOO_BIG: u8 = 24;
 const ERROR_ID_ZERO_QUANTITY: u8 = 28;
-const ERROR_ID_CANNOT_CHANGE_WCCC_ASSET_SCHEME: u8 = 29;
 const ERROR_ID_DISABLED_TRANSACTION: u8 = 30;
 const ERROR_ID_INVALID_SIGNER_OF_WRAP_CCC: u8 = 31;
```
We have some remaining error ids. Let's remove them.

sgkim126 commented:
It seems that this commit removes a shard itself not only assets. Is it right?
IMO, it's better to keep shards and shard manipulation transactions like change owners, create shards, etc.

majecty commented:
@sgkim126 I removed shard because I removed all shard transactions. I tried to inform you in Gitter, before removing shards, but it seems not delivered to you. I'll care about communication more.

majecty commented:
I'll restore Shard code back.

sgkim126 commented:
> @sgkim126 I removed shard because I removed all shard transactions. I tried to inform you in Gitter, before removing shards, but it seems not delivered to you. I'll care about communication more.

What I understand was you would remove all transactions that change the states of shards. I reread your message. It was my confusion.

majecty commented:
I'm creating small PRs that only removes the asset related code.
So we don't need this PR anymore.
